onnxruntime/contrib_ops/cuda/bert/packed_multihead_attention.cc

```diff
@@ -0,0 +1,267 @@
+
+  scale_ = info.GetAttrOrDefault<float>("scale", 0.0f);
+
+#if USE_FLASH_ATTENTION
+  disable_memory_efficient_attention_ = onnxruntime::ParseEnvironmentVariableWithDefault<bool>(
+      attention::kDisableMemoryEfficientAttention, false);
+#else
+  disable_memory_efficient_attention_ = true;
+#endif
+}
+
+template <typename T>
+Status PackedMultiHeadAttention<T>::CheckInputs(const TensorShape& query_shape,
+                                                const Tensor* key,
+                                                const Tensor* value,
+                                                const TensorShape& token_offset_shape,
+                                                const TensorShape& cu_seq_len_shape,
+                                                const Tensor* relative_position_bias,
+                                                PackedAttentionParameters& parameters) const {
+  // Shapes of inputs and output:
+  // When Q, K and V are not packed:
+  //   Input 'query':                      (token_count, hidden_size)
+  //   Input 'key':                        (token_count, hidden_size)
+  //   Input 'value':                      (token_count, v_hidden_size)
+  // When Q, K and V are packed:
+  //   Input 'query':                      (token_count, num_heads, 3, head_size)
+  //   Input 'key':                        None
+  //   Input 'value':                      None
+  // Input 'token_offset':                 (batch_size, sequence_length)
+  // Input 'cumulative_sequence_length':   (batch_size + 1)
+  // Input 'relative_position_bias':       (batch_size or 1, num_heads, sequence_length, sequence_length) or None
+  // Output 'output':                      (token_count, v_hidden_size)
+
+  const auto& query_dims = query_shape.GetDims();
+  if (query_dims.size() != 2 && query_dims.size() != 4) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                           "Input 'query' is expected to have 2 or 4 dimensions in packing mode, got ",
+                           query_dims.size());
+  }
+  int64_t token_count = query_dims[0];
+  int64_t hidden_size = (query_dims.size() == 2) ? query_dims[1] : (query_dims[1] * query_dims[3]);
+
+  const auto& token_offset_dims = token_offset_shape.GetDims();
+  if (token_offset_dims.size() != 2) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                           "Input 'token_offset' is expected to have 2 dimensions in packing mode, got ",
+                           token_offset_dims.size());
+  }
+
+  int64_t batch_size = token_offset_dims[0];
+  int64_t sequence_length = token_offset_dims[1];
+
+  int64_t v_hidden_size = hidden_size;
+  if (query_dims.size() == 4) {
+    if (key != nullptr || value != nullptr) {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                             "Input 'key' and 'value' is expected to be empty when 'query' has 4 dimensions in packing mode");
+    }
+  } else {  // query_dims.size() == 2
```

Comment:
assert this? (query_dims.size() == 2)

Reply:
Not needed since we've checked query_dims is either 2 or 4 in the code above.